File: src/dotnet/PolyBus/Transport/Transactions/Messages/OutgoingMessage.cs
using System.Diagnostics;

namespace PolyBus.Transport.Transactions.Messages;

[DebuggerStepThrough]
public class OutgoingMessage(IPolyBus bus, object message, string? endpoint = null, MessageInfo? messageInfo = null) : Message(bus)
{
    /// <summary>
    /// If the transport supports delayed messages, this is the time at which the message should be delivered.
    /// </summary>
    public virtual DateTime? DeliverAt { get; set; }

    /// <summary>
    /// The message info describing metadata about the message.
    /// </summary>
    public virtual MessageInfo MessageInfo { get; set; } = messageInfo ?? bus.Messages.GetMessageInfo(message.GetType());

    /// <summary>
    /// The type of the message.
    /// </summary>
    public virtual Type MessageType { get; set; } = message.GetType();

    /// <summary>
    /// An optional location to explicitly send the message to.
    /// </summary>
    public virtual string? Endpoint { get; set; } = endpoint;

    /// <summary>
    /// The serialized message body contents.
    /// </summary>
    public virtual string Body { get; set; } = string.Empty;

    /// <summary>
    /// The message object.
    /// </summary>
    public virtual object Message { get; set; } = message;
}

File: src/kotlin/src/main/kotlin/polybus/transport/transactions/messages/OutgoingMessage.kt
package polybus.transport.transactions.messages

import polybus.IPolyBus
import java.time.Instant

open class OutgoingMessage(
    bus: IPolyBus,
    message: Any,
    endpoint: String? = null,
    messageInfo: MessageInfoValue? = null
) : Message(bus) {
    open var deliverAt: Instant? = null
    open var messageInfo: MessageInfoValue = messageInfo ?: bus.messages.getMessageInfo(message.javaClass)
    open var messageType: Class<*> = message.javaClass
    open var endpoint: String? = endpoint
    open var body: String = ""
    open var message: Any = message
}
